multicore-hello.c

```diff
@@ -30,14 +30,25 @@ int secondary_main(void) {
 	if(hartid == 0) {
 		int rc = metal_lock_init(&my_lock);
 		if(rc != 0) {
-			printf("Failed to initialize my_lock: %d", rc);
+			puts("Failed to initialize my_lock\n");
 			exit(1);
 		}
 
+		/* Ensure that the lock is initialized before any readers of
+		 * _start_other */
+		__asm__ ("fence rw,w"); /* Release semantics */
+
 		_start_other = 1;
 
 		return main();
 	} else {
+		/* We can't count on hart 0 to initialize _start_other before
+		 * the other harts get here because hart 0 is in charge of copying
+		 * the data section, zeroing bss, and running the constructors.
+		 * The other harts will always zero _start_other here long before
+		 * hart 0 gets around to setting it to 1 */
+		_start_other = 0;
```

Comment:
I thought that this was done by initializing "volatile int _start_other = 0;"?
I worry that this is a non-deterministic race, and we are relying on harts >0 winning the race before hart 0.

Reply:
We are relying on that, but hart 0 has a lot of work to do before it gets here that the other harts don't. I expect it's something we can get away with in the short-term so I created a tracking issue in Freedom Metal to note that we need a better solution.